actionview/lib/action_view/helpers/asset_tag_helper.rb

```diff
@@ -521,13 +521,40 @@ def resolve_link_as(extname, mime_type)
           end
         end
 
-        def send_preload_links_header(preload_links)
+        MAX_HEADER_SIZE = 8_000 # Some HTTP client and proxies have a 8kiB header limit
+        def send_preload_links_header(preload_links, max_header_size: 8_000)
           if respond_to?(:request) && request
             request.send_early_hints("Link" => preload_links.join("\n"))
           end
 
           if respond_to?(:response) && response
-            response.headers["Link"] = [response.headers["Link"].presence, *preload_links].compact.join(",")
+            header = response.headers["Link"]
+            header = header ? header.dup : +""
+
+            # rindex count characters not bytes, but we assume non-ascii characters
+            # are rare in urls, and we have a 192 bytes margin.
+            last_line_offset = header.rindex("\n")
+            last_line_size = if last_line_offset
+              header.bytesize - last_line_offset
+            else
+              header.bytesize
+            end
+
+            preload_links.each do |link|
+              if link.bytesize + last_line_size + 1 < MAX_HEADER_SIZE
+                unless header.empty?
+                  header << ","
+                  last_line_size += 1
+                end
+              else
+                header << "\n"
+                last_line_size = 0
+              end
+              header << link
+              last_line_size += link.bytesize
+            end
+
+            response.headers["Link"] = header
```

Comment:
I experimented with tidying up while preserving performance.  What do you think about the following?

```suggestion
            old_header = response.headers["Link"] || ""
            header = [old_header, *preload_links].join(",")

            bytesize = old_header.bytesize
            i = old_header.length
            preload_links.each do |link|
              bytesize += link.bytesize + 1
              if bytesize > MAX_HEADER_SIZE
                header[i] = "\n"
                bytesize = link.bytesize
              end
              i += link.length + 1
            end

            response.headers["Link"] = header
```


<details>
  <summary>Benchmark script</summary>
  
  ```ruby
  # frozen_string_literal: true
  require "benchmark/ips"

  MAX_HEADER_SIZE = 8_000
  PRELOAD_LINKS = 1000.times.map { |i| "http://example.com/style.css?#{i}" }

  def shovel(old_header, preload_links)
    header = old_header.dup

    # rindex count characters not bytes, but we assume non-ascii characters
    # are rare in urls, and we have a 192 bytes margin.
    last_line_offset = header.rindex("\n")
    last_line_size = if last_line_offset
      header.bytesize - last_line_offset
    else
      header.bytesize
    end

    preload_links.each do |link|
      if link.bytesize + last_line_size + 1 < MAX_HEADER_SIZE
        unless header.empty?
          header << ","
          last_line_size += 1
        end
      else
        header << "\n"
        last_line_size = 0
      end
      header << link
      last_line_size += link.bytesize
    end

    header
  end

  def post_hoc(old_header, preload_links)
    header = [old_header, *preload_links].join(",")

    bytesize = old_header.bytesize
    i = old_header.length
    preload_links.each do |link|
      bytesize += link.bytesize + 1
      if bytesize > MAX_HEADER_SIZE
        header[i] = "\n"
        bytesize = link.bytesize
      end
      i += link.length + 1
    end

    header
  end

  Benchmark.ips do |x|
    x.report("shovel") { shovel(PRELOAD_LINKS.first, PRELOAD_LINKS) }
    x.report("post-hoc") { post_hoc(PRELOAD_LINKS.first, PRELOAD_LINKS) }
    x.compare!
  end
  ```
</details>


```
Comparison:
            post-hoc:     2461.1 i/s
              shovel:     1879.2 i/s - 1.31x  (± 0.00) slower
```

It's a bit slower when non-ASCII characters are involved, though.

Reply:
Hum, that's interesting. I'm a bit worried about the extra complexity of having to deal with the bytes / index translation (even though most preload links would be ascii only).

I'm also not sure the extra performance would hold if the existing header is already large.

Reply:
Actually it does even better in these case. I'll update.

Reply:
Hum, turns out the code isn't quite equivalent. Since your's don't look for a `\n` in the existing header, it does generate much more lines that mine.

I'm not sure which is best to be honest. Headers are not compressed (except in HTTP2, but even there I think it's limited to a static dictionary), so not sure wether limiting the number of headers is important or not.

I wasn't expecting over 8KiB of links in the first place though, and I think it's quite a fringe use case, so maybe we should just not bother and just not try to batch links?

Reply:
> Hum, turns out the code isn't quite equivalent. Since your's don't look for a `\n` in the existing header, it does generate much more lines that mine.

Ah, I was thinking `send_preload_links_header` was only called once, so it would be one extra line at most, but I see it is called multiple times.  We could adjust the starting `bytesize` to `old_header[old_header.rindex("\n")&.succ..].bytesize`.  It is less pretty, but it doesn't seem to adversely affect performance (thanks to Ruby's string tail sharing):

<details>
  <summary>Benchmark script</summary>

  ```ruby
  # frozen_string_literal: true
  require "benchmark/ips"

  MAX_HEADER_SIZE = 8_000
  PRELOAD_LINKS = 1000.times.map { |i| "http://example.com/style.css?#{i}" }
  OLD_HEADER = "#{"1" * MAX_HEADER_SIZE}\n#{"2" * (MAX_HEADER_SIZE / 2)}"

  def shovel(old_header, preload_links)
    header = old_header.dup

    # rindex count characters not bytes, but we assume non-ascii characters
    # are rare in urls, and we have a 192 bytes margin.
    last_line_offset = header.rindex("\n")
    last_line_size = if last_line_offset
      header.bytesize - last_line_offset
    else
      header.bytesize
    end

    preload_links.each do |link|
      if link.bytesize + last_line_size + 1 < MAX_HEADER_SIZE
        unless header.empty?
          header << ","
          last_line_size += 1
        end
      else
        header << "\n"
        last_line_size = 0
      end
      header << link
      last_line_size += link.bytesize
    end

    header
  end

  def post_hoc(old_header, preload_links)
    header = [*old_header, *preload_links].join(",")

    # bytesize = old_header.bytesize
    bytesize = old_header[old_header.rindex("\n")&.succ..].bytesize
    i = old_header.length
    preload_links.each do |link|
      bytesize += link.bytesize + 1
      if bytesize > MAX_HEADER_SIZE
        header[i] = "\n"
        bytesize = link.bytesize
      end
      i += link.length + 1
    end

    header
  end

  Benchmark.ips do |x|
    x.report("shovel") { shovel(OLD_HEADER, PRELOAD_LINKS) }
    x.report("post-hoc") { post_hoc(OLD_HEADER, PRELOAD_LINKS) }
    x.compare!
  end
  ```
</details>


```
Comparison:
            post-hoc:     2463.7 i/s
              shovel:     1845.6 i/s - 1.33x  (± 0.00) slower
```

(Note: I also changed the starting header to be one and a half maxed-out lines.)



> I'm not sure which is best to be honest. Headers are not compressed (except in HTTP2, but even there I think it's limited to a static dictionary), so not sure wether limiting the number of headers is important or not.

The `Link` header *is* in the [HPACK static dictionary](https://tools.ietf.org/html/rfc7541#appendix-A), so perhaps it does not make that much of a difference, as long as you are using HTTP2.